### PR TITLE
add a note on numerical accuracy

### LINKED ISSUE #54437: Improve numeric stability notes
## 📚 Documentation

At the triage review meeting, @ngimel pointed out that we get issues of the "f(x[0]) != f(x)[0]" variety quite often (where f is some function that operates on batches independently) (e.g. https://github.com/pytorch/pytorch/issues/54369). The reason why those are not always equal is because of floating point error. 

We should try to add a note about this into our documentation that we can reference. One potential section to add it is in the reproducibility note (https://pytorch.org/docs/stable/notes/randomness.html) but alternatively we can add a separate note for numeric stability.

### PR DESCRIPTION
**Pull request by ngimel**
Per title
Fixes #54437

### CODE REVIEW

File: docs/source/notes/numerical_accuracy.rst [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+------------------------------------------
+
+Many operations in pytorch support batched computation, where the same operation is performed
+for the elements of the batches of inputs. An example of this is :meth:`torch.mm` and
+:meth:`torch.bmm`. It is possible to implement batched computation as a loop over batch elements,
+and apply the necessary math operations to the individual batch elements, for efficiency reasons
+we are not doing that, and typically perform computation for the whole batch. The mathematical
+libraries that we are calling, and pytorch internal implementations of operations can produces
+slightly different results in this case, compared to non-batched computations. In particular,
+let ``A`` and ``B`` be 3D tensors with the dimensions suitable for batched matrix multiplication.
```
Comment (albanD) [inline docs/source/notes/numerical_accuracy.rst:23]: Do we want to make B a vector here to avoid having to index it below?

File: docs/source/notes/numerical_accuracy.rst [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+Then ``A@B[0]`` (the first element of the batched result) is not guaranteed to be bitwise
+identical to ``A[0]@B[0]`` (the matrix product of the first elements of the input batches)
+even though mathematically it's an identical computation.
```
Comment (albanD) [inline docs/source/notes/numerical_accuracy.rst:26]: Do we want to have a code sample and show an example difference here?

File: docs/source/notes/numerical_accuracy.rst [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+``A[:,0].sum()``.
+
+Extremal values
+---------------
+
+When inputs contain large values such that intermediate results may overflow the range of the
+used datatype, the end result may overflow too, even though it is representable in the original
+datatype. E.g.::
+
+    import torch
```
Comment (albanD) [inline docs/source/notes/numerical_accuracy.rst:40]: I think you want to use `..code::` and `>>>` for the rendering and copying to be correct no?

File: docs/source/notes/numerical_accuracy.rst [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    a=torch.tensor([1e20, 1e20]) #fp32 type by default
+    a.norm() # produces tensor(inf)
+    a.double().norm() #produces tensor(1.4142e+20, dtype=torch.float64), representable in fp32
+
+TensorFloat-32(TF32) on Nvidia Ampere devices
+---------------------------------------------
+
+On Ampere Nvidia GPUs, pytorch by default uses TensorFloat32 (TF32) to speed up mathematically
+intensive operations, in particular matrix multiplications and gemms. When operation is performed
```
Comment (albanD) [inline docs/source/notes/numerical_accuracy.rst:49]: aren't gemms matrix multiplication? Is there an important distinction you want to make here.

File: docs/source/notes/numerical_accuracy.rst [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+using TF32 tensor cores, only first 10 bits of input mantissa are read. This leads to less accurate
```
Comment (albanD) [inline docs/source/notes/numerical_accuracy.rst:50]: only the first .. the input ?

File: docs/source/notes/numerical_accuracy.rst [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+results, and surprising results such as multiplying a matrix by identity matrix produces
+results that are different from the input.
+Most neural network workloads have the same convergence behavior when using tf32 as they have
+with fp32, however, if better accuracy is desired, TF32 can be turned off.
```
Comment (albanD) [inline docs/source/notes/numerical_accuracy.rst:54]: Do you want to link to the function that turns that off here? Or it is too specific?

**Review by albanD — COMMENTED** [review]
That looks quite good. Mostly small details below.

One thing I'm wondering is if we should add a couple of basic stuff at the beginning like the fact that float has 6 digit of precision and so anything below is noise or ops between numbers that are too different is wrong.
Or instead of writing these basic stuff, reference good material written somewhere else on that topic?

**Review by albanD — APPROVED** [review]
It reads very well now. Thanks for the update!

**Review by albanD — APPROVED** [review]
final rev
